### PR TITLE
Add custom JsonConverter for BinaryData that converts to Base64 string

### LINKED ISSUE #70267: Consider adding JsonConverter for BinaryData that converts to Base64 string
Currently any models that are serialized with System.Text.Json that contain a [BinaryData ](https://docs.microsoft.com/en-us/dotnet/api/system.binarydata?view=dotnet-plat-ext-6.0) property will not have the data populated on the resulting JSON. This is because BinaryData doesn't contain any public properties.

One potential solution would be to define a JsonConverter that converts the BinaryData bytes to a base64 string and includes this in the JSON. For deserialization, the BinaryData JSON is assumed to be a Base64 string:

```c#
    internal class BinaryDataConverter : JsonConverter<BinaryData>
    {
        public override BinaryData Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            return BinaryData.FromBytes(reader.GetBytesFromBase64());
        }

        public override void Write(Utf8JsonWriter writer, BinaryData value, JsonSerializerOptions options)
        {
            writer.WriteBase64StringValue(value.ToMemory().Span);
        }
    }
```
The converter would then be added to the BinaryData type:

```c#
    [JsonConverter(typeof(BinaryDataConverter))]
    public class BinaryData {
```
This would match the behavior of how byte arrays get serialized with System.Text.Json - they are [converted to Base64](https://github.com/dotnet/runtime/blob/4a4e347b964a2c8d2216ec382e4fb481965bb2fc/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/ByteArrayConverter.cs). For byte arrays, it looks like the ByteArrayConverter is defined as a default converter in System.Text.Json. This wouldn't work for BinaryData as BinaryData depends on System.Text.Json, so that would introduce a circular dependency. Instead, I think we'd need to add the converter directly to the BinaryData type.

### PR DESCRIPTION
**Pull request by ovidiucosteanet**
- Created custom converter as per described in the issue
- Added the converter also to the ref BinaryData class
- Added appropriate tests for serialization and deserialization of a class that has a member of BinaryData type

Fixes #70267

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73448) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73448) before we can accept your contribution.<br/><br/>:x: ovidiucosteanet [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=73448)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=73448) it.</sub>

File: src/libraries/System.Memory.Data/ref/System.Memory.Data.cs [hunk 1]
```diff
@@ -37,4 +38,10 @@ public partial class BinaryData
         public System.IO.Stream ToStream() { throw null; }
         public override string ToString() { throw null; }
     }
+
+    public sealed class BinaryDataConverter : System.Text.Json.Serialization.JsonConverter<BinaryData>
```
Comment (JoshLove-msft) [inline src/libraries/System.Memory.Data/ref/System.Memory.Data.cs:42]: Should this be public?

Comment (ovidiucosteanet) [reply src/libraries/System.Memory.Data/ref/System.Memory.Data.cs:42]: Before adding it to ref assemblies it was internal. Not being familiar with ref assemblies and seeing that other assemblies in the repo have no internal types, but rather public types with internal constructor, I tried to do the same but got  `System.InvalidOperationException : The converter specified on 'System.BinaryData' does not derive from JsonConverter or have a public parameterless constructor.` I just checked [Microsoft Docs - Ref assemblies](https://docs.microsoft.com/en-us/dotnet/standard/assembly/reference-assemblies) and it seems it should be fine to make the type itself internal.

File: src/libraries/System.Memory.Data/src/System.Memory.Data.csproj [hunk 2]
```diff
@@ -9,14 +9,12 @@ Commonly Used Types:
 System.BinaryData</PackageDescription>
   </PropertyGroup>
 
-  <ItemGroup>
+  <ItemGroup>
```
Comment (jsquire) [inline src/libraries/System.Memory.Data/src/System.Memory.Data.csproj:12]:
```suggestion
  <ItemGroup>
```

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you for your contribution @ovidiucosteanet !

**Comment by ovidiucosteanet** [comment]
Thank you @adamsitnik! Is there anything else that I should do regarding this PR? As I can see I do not have merge rights.

File: src/libraries/System.Memory.Data/src/System/BinaryDataConverter.cs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace System
+{
+    internal sealed class BinaryDataConverter : JsonConverter<BinaryData>
+    {
+        public sealed override BinaryData? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
```
Comment (krwq) [inline src/libraries/System.Memory.Data/src/System/BinaryDataConverter.cs:11]: consider supporting null tokens and overriding https://docs.microsoft.com/en-us/dotnet/api/system.text.json.serialization.jsonconverter-1.handlenull?view=net-6.0

**Review by krwq — APPROVED** [review]
LGTM

File: src/libraries/System.Memory.Data/src/System/BinaryDataConverter.cs (hunk 3 above)
Comment (krwq) [reply src/libraries/System.Memory.Data/src/System/BinaryDataConverter.cs:11]: (assuming you want i.e. serialize them to empty binary data)

Comment (krwq) [reply src/libraries/System.Memory.Data/src/System/BinaryDataConverter.cs:11]: on the second thought that might be unexpected for users 😄

**Pull request merged**